### PR TITLE
Incremental suffix to cloned entities's name (issue #575)

### PR DESCRIPTION
**Pull request by AleziaKurdis**
This enhancement adds an incremental suffix to the name 
of the cloned entities resulting from "Duplicate" and "Paste".
The goal is to help to figure which entities is the coned one.
This addresses issue kasenvr#575

### CODE REVIEW

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js [hunk 1]
```diff
@@ -31,6 +33,55 @@ function deepCopy(v) {
...
+    var rippedOriginalName = removeNameAppendice(originalName);
+
+    do {
+        duplicateNumber++;
+                if (rippedOriginalName === "") {
+                    duplicateName = "(" + duplicateNumber + ")";
+                } else {
+                    duplicateName = rippedOriginalName + " (" + duplicateNumber + ")";
+                }
+                existingNames = Entities.findEntitiesByName(duplicateName, Vec3.ZERO, ENTIRE_DOMAIN_SCAN_RADIUS, false);
```
Comment (HifiExperiments) [inline scripts/system/create/entitySelectionTool/entitySelectionTool.js:49]: I think this is at the core of why I don’t love this solution.  1. it’s not a good idea to call this method in a loop (it can be expensive, it’s blocking, and this loop could be huge), 2. what if you’ve got Box, Box (2), and Box (3), and you delete Box (2), and then copy Box again?

Comment (AleziaKurdis) [reply scripts/system/create/entitySelectionTool/entitySelectionTool.js:49]: What other solution do we have?
a sequence per name kept for the session?

Or simply not doing this functionality.
For sure, this can rapidly become an annoying feature when used for repetitive cloning.

Comment (HifiExperiments) [reply scripts/system/create/entitySelectionTool/entitySelectionTool.js:49]: keeping track of it per name won’t work because the script could be reloaded or stopped

like I said in the issue, I would vote to have Box always become Box (2), even if you clone it multiple times

**Comment by AleziaKurdis** [comment]
Just pushed a simpler solution without any hazardous uniqueness check.
Cloned entities will systematically get the suffix " (2)", unless their name has already a suffix. 
In that case, we simply increment the suffix.

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js [hunk 2]
```diff
@@ -485,6 +528,7 @@ SelectionManager = (function() {
             } else {
                 delete properties.position;
             }
+            properties.name = getDuplicateAppendedName(properties.name);
```
Comment (HifiExperiments) [inline scripts/system/create/entitySelectionTool/entitySelectionTool.js:531]: what's the difference between pasting and duplicating?  do we want only duplicating to do this new logic?

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js [hunk 3]
```diff
@@ -31,6 +32,46 @@ function deepCopy(v) {
...
+                        rippedName = name.substr(0, i-1);
+                        existingSequence = parseInt(sequenceReader);
+                        break;
+                    } else {
+                        sequenceReader = currentChar + sequenceReader;
+                    }
+                }
+            }
+        }
+    }
```
Comment (HifiExperiments) [inline scripts/system/create/entitySelectionTool/entitySelectionTool.js:65]: this logic *seems* sound to me, but it's a little hard to follow, which means it will also be a little hard to debug if something ever goes wrong.

I wonder if it would be easier to use regexs?  I don't know what JS regex support looks like

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js (hunk 2 above)
Comment (AleziaKurdis) [reply scripts/system/create/entitySelectionTool/entitySelectionTool.js:531]: Duplicate is Move + CTRL... generating a clone of the entity. (this one is right I think)

But just realize that it was not a good implementation for "Paste" (when you do Copy > Paste) because you might also have done "Cut > Paste" and this is going to be wrong in that second case. 
We don't want a cut-paste to increment...  (I would need to check that.)

I envision that more for the "Duplicate" since it's what I do 100% of the time when I use this.
It's probably not too difficult to address Copy versus Cut.

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js (hunk 3 above)
Comment (AleziaKurdis) [reply scripts/system/create/entitySelectionTool/entitySelectionTool.js:65]: I've simply used something I trust... regex is just not intuitive "for me" and I don't know what surprise it could bring.
I will check that option.

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js (hunk 2 above)
Comment (AleziaKurdis) [reply scripts/system/create/entitySelectionTool/entitySelectionTool.js:531]: I will check for that

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js (hunk 3 above)
Comment (HifiExperiments) [reply scripts/system/create/entitySelectionTool/entitySelectionTool.js:65]: yeah not super important, just wanted to mention it in case it would be easier

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js (hunk 2 above)
Comment (HifiExperiments) [reply scripts/system/create/entitySelectionTool/entitySelectionTool.js:531]: could you also copy from another domain and paste in a new one?

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-024980e_91ebd378.dmg

**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-024980e_91ebd378.sh
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-024980e.tar_91ebd378.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-024980e.tar_91ebd378.gz

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-024980e_91ebd378.exe

File: scripts/system/create/entitySelectionTool/entitySelectionTool.js (hunk 2 above)
Comment (AleziaKurdis) [reply scripts/system/create/entitySelectionTool/entitySelectionTool.js:531]: "Duplicate" makes sense to get the suffix (and have it incremented)

"Paste" doesn't. It's not possible for the script to be sure of the origin of what is in the clipboard.
It would not make sense to add the suffix if you paste something copied on another domain.
It would not  make sense to add the suffix to something we cut and then paste (to let say move it somewhere else  in the same or in another domain)
we can paste the cut object N time. If we can track this for the current run, it's impossible the second we close and reopen the creat app.

I will adjust the fix to have this only for "Duplicate".
No suffix for pasted entities.

**Comment by vircadia-build-notifier** [comment]
The following links are available: 
**build (ubuntu-18.04, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-ee5de58.tar_2cd1fc84.Z
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-ee5de58.tar_2cd1fc84.gz
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-ee5de58_2cd1fc84.sh

**build (windows-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-ee5de58_2cd1fc84.exe

**build (macOS-latest, full)**
 - https://athena-public.s3.eu-west-3.amazonaws.com/builds/PR/kasenvr_project-athena/PR603/Vircadia-Alpha-PR603-ee5de58_2cd1fc84.dmg

**Comment by kasenvr** [comment]
image
Appears to be working as expected for me. :)
